### PR TITLE
v1.1: record-sales feature

### PR DESCRIPTION
**Pull request by HyperionNKJ**
1) Modified UG and DG to complement with other components.
2) Users can now record sales volume of menu items (UI window's panel is set to show person list instead of sales records list because doing the latter will fail test cases. All test cases pass as of now.) 
3) Testing is now completed except for UI portion.

### CODE REVIEW

**Comment by AZhiKai** [comment]
I see that your CI are failing due to EOF. 

Take a look at this post: https://stackoverflow.com/questions/16761227/how-to-make-intellij-idea-insert-a-new-line-at-every-end-of-file

**Comment by HyperionNKJ** [comment]
PR ready for review

File: src/main/java/seedu/address/model/UserPrefs.java [hunk 1]
```diff
@@ -12,7 +12,7 @@
 public class UserPrefs {
 
     private GuiSettings guiSettings;
-    private Path addressBookFilePath = Paths.get("data", "addressbook.xml");
+    private Path addressBookFilePath = Paths.get("data" , "addressbook.xml");
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/UserPrefs.java]: Is there really a need for a space here?

File: src/main/java/seedu/address/model/salesrecord/Date.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * Represents a date in the calendar in DD-MM-YYYY format Guarantees: immutable; is valid as declared in {@link
+ * #isValidDate(String)}
+ */
+public class Date {
+
+    public static final String MESSAGE_DATE_CONSTRAINTS =
+            "Dates should only contain numbers, and it should be in the format DD-MM-YYYY.\nThe date must exist in "
+                    + "the calendar";
+    public final String date;
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/salesrecord/Date.java]: Should have a space in between.

File: src/main/java/seedu/address/model/salesrecord/Date.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    @Override
+    public String toString() {
+        return date;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof Date // instanceof handles nulls
+                && date.equals(((Date) other).date)); // state check
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/salesrecord/Date.java]: Can further indent `&&` under "(other...) like this:

```
return other == this // short circuit if same object
                || (other instanceof Date // instanceof handles nulls
                   && date.equals(((Date) other).date)); // state check
```

File: src/main/java/seedu/address/model/salesrecord/Price.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    @Override
+    public String toString() {
+        return String.valueOf(value);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof Price // instanceof handles nulls
+                && value == ((Price) other).value); // state check
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/salesrecord/Price.java]: Same as earlier, can indent the `&&`.

File: src/main/java/seedu/address/model/salesrecord/QuantitySold.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    @Override
+    public String toString() {
+        return String.valueOf(value);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof QuantitySold // instanceof handles nulls
+                && value == ((QuantitySold) other).value); // state check
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/salesrecord/QuantitySold.java]: Same as earlier, can indent the `&&`.

File: src/main/java/seedu/address/model/salesrecord/exceptions/RecordNotFoundException.java [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package seedu.address.model.salesrecord.exceptions;
+
+/**
+ * Signals that the operation is unable to find the specified record.
+ */
+public class RecordNotFoundException extends RuntimeException {}
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/salesrecord/exceptions/RecordNotFoundException.java]: "Record" is quite vague. What kind of record? 
In general, my suggestion is to rename `SalesRecord` as just `Sales`. So here you can list as `SalesNotFoundException`.

File: src/main/java/seedu/address/storage/XmlAdaptedRecord.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    /**
+     * Converts a given SalesRecord into this class for JAXB use.
+     *
+     * @param source future changes to this will not affect the created XmlAdaptedRecord
+     */
+    public XmlAdaptedRecord(SalesRecord source) {
+        date = source.getDate().date;
+        itemName = source.getName().fullName;
```
Comment (AZhiKai) [inline src/main/java/seedu/address/storage/XmlAdaptedRecord.java]: It seems like you're accessing `fullName`...which implies it is a public variable? You should encapsulate the object's field and return it using `toString()`. 

So your `toString()` can be returning `return fullName;`.

File: src/main/java/seedu/address/model/salesrecord/Price.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents the price of a menu item sold in the sales record
+ * Guarantees: immutable; is valid as declared in {@link #isValidPrice(String)}
+ */
+public class Price {
+
+    public static final String MESSAGE_PRICE_CONSTRAINTS =
+            "Price should be a positive number in SGD.";
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/salesrecord/Price.java]: Perhaps remove the "SGD". We should consider for if the system is used in other countries of other currencies. Sometimes being too specific isn't good.

File: src/main/java/seedu/address/model/salesrecord/ItemName.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    @Override
+    public String toString() {
+        return fullName;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof seedu.address.model.salesrecord.ItemName // instanceof handles nulls
+                && fullName.equals(((seedu.address.model.salesrecord.ItemName) other).fullName)); // state check
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/salesrecord/ItemName.java]: Same as earlier, can indent the `&&`.

File: src/main/java/seedu/address/model/salesrecord/ItemName.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    public static final String MESSAGE_NAME_CONSTRAINTS =
+            "Menu item names should only contain alphanumeric characters and spaces, and it should not be blank";
+
+    /*
+     * The first character of the address must not be a whitespace,
+     * otherwise " " (a blank string) becomes a valid input.
+     */
+    public static final String NAME_VALIDATION_REGEX = "[\\p{Alnum}][\\p{Alnum} ]*";
+
+    public final String fullName;
```
Comment (AZhiKai) [inline src/main/java/seedu/address/model/salesrecord/ItemName.java]: Privatize `fullName`. To access it, do `...getName().toString()`.

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 11]
```diff
@@ -21,8 +21,12 @@
 import seedu.address.logic.commands.SelectCommand;
 import seedu.address.logic.commands.UndoCommand;
 import seedu.address.logic.commands.accounts.CreateCommand;
+import seedu.address.logic.commands.salescommands.RecordSalesCommand;
+
 import seedu.address.logic.parser.accounts.CreateCommandParser;
 import seedu.address.logic.parser.exceptions.ParseException;
+import seedu.address.logic.parser.salescommandsparser.RecordSalesCommandParser;
+
```
Comment (AZhiKai) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java]: There seems to be a redundant empty line here.

File: src/main/java/seedu/address/logic/commands/salescommands/RecordSalesCommand.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+        model.addRecord(toAdd);
+        model.commitAddressBook();
+        return new CommandResult(String.format(MESSAGE_RECORD_SALES_SUCCESS, toAdd));
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof RecordSalesCommand // instanceof handles nulls
+                && toAdd.equals(((RecordSalesCommand) other).toAdd));
```
Comment (AZhiKai) [inline src/main/java/seedu/address/logic/commands/salescommands/RecordSalesCommand.java]: Same as earlier, can indent the `&&`.

File: src/main/java/seedu/address/commons/events/ui/RecordPanelSelectionChangedEvent.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    public RecordPanelSelectionChangedEvent(SalesRecord newSelection) {
+        this.newSelection = newSelection;
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName();
+    }
+
+    public SalesRecord getNewSelection() {
```
Comment (AZhiKai) [inline src/main/java/seedu/address/commons/events/ui/RecordPanelSelectionChangedEvent.java]: Missing Javadoc.

File: src/main/java/seedu/address/logic/Logic.java [hunk 14]
```diff
@@ -26,6 +27,10 @@
     /** Returns an unmodifiable view of the filtered list of persons */
     ObservableList<Account> getFilteredAccountList();
 
+    /** Returns an unmodifiable view of the filtered list of sales records */
+    ObservableList<SalesRecord> getFilteredRecordList();
+
+
```
Comment (AZhiKai) [inline src/main/java/seedu/address/logic/Logic.java]: Seems to have a redundant empty line here.

File: src/main/java/seedu/address/storage/XmlAdaptedRecord.java [hunk 15]
```diff
@@ -0,0 +1,148 @@
+package seedu.address.storage;
```
Comment (AZhiKai) [inline src/main/java/seedu/address/storage/XmlAdaptedRecord.java]: You should put this file inside the `elements` package.

File: src/main/java/seedu/address/storage/XmlSerializableAddressBook.java [hunk 16]
```diff
@@ -65,13 +86,29 @@ public void processPersons() throws IllegalValueException {
...
+    }
+
     /**
      * Converts this account record into the model's {@code Account} object.
      *
      * @throws IllegalValueException if there were any data constraints violated or duplicates in the {@code
      *         XmlAdaptedAccount}.
      */
-    public void processAccounts() throws IllegalValueException {
+    public void accountToModelType() throws IllegalValueException {
```
Comment (AZhiKai) [inline src/main/java/seedu/address/storage/XmlSerializableAddressBook.java]: Perhaps you can help me privatize this method as well? :)

**Comment by HyperionNKJ** [comment]
Thank you for your review @AZhiKai  I have just merged changes from upstream's master into this PR.
Will work on issue #61 and #62 in a new separate PR.

**Comment by AZhiKai** [comment]
> Thank you for your review @AZhiKai I have just merged changes from upstream's master into this PR.
> Will work on issue #61 and #62 in a new separate PR.

Sounds good! 😄

**Comment by AZhiKai** [comment]
Hi @HyperionNKJ , please update the branch first before we proceed to merging your PR.

**Pull request merged**